src/vs/workbench/contrib/terminal/browser/xterm/decorationAddon.ts

```diff
@@ -167,12 +169,25 @@ export class DecorationAddon extends Disposable implements ITerminalAddon {
 		for (const command of capability.commands) {
 			this.registerCommandDecoration(command);
 		}
-		this._commandFinishedListener = capability.onCommandFinished(command => {
-			if (command.command.trim().toLowerCase() === 'clear' || command.command.trim().toLowerCase() === 'cls') {
-				this.clearDecorations();
-				return;
+		this._commandFinishedListener = capability.onCommandFinished(command => this.registerCommandDecoration(command));
+	}
+
+	private _addCommandClearedListener(): void {
+		if (this._commandStartedListener) {
+			return;
+		}
+		const capability = this._capabilities.get(TerminalCapability.CommandDetection);
+		if (!capability) {
+			return;
+		}
+
+		this._commandClearedListener = capability.onCommandCleared(commands => {
+			for (const value of this._decorations.values()) {
+				if (commands.find(c => value.decoration.marker === c.marker)) {
+					value.decoration.dispose();
+					dispose(value.disposables);
+				}
 			}
```

Comment:
Marker id is the key so this can be much faster, something like this:

```suggestion
		for (const command of commands) {
			const match = this._decorations.get(command.marker.id);
			if (match) {
				match.decoration.dispose();
				dispose(match.disposables);
			}
		}
```